### ISSUE TITLE
setModelHandling doesn't work with already created vehicles

### ISSUE THREAD

**Issue by ArranTuna** [multitheftauto/mtasa-blue#553]
**Describe the bug**
"setModelHandling" function doesn't affect already created vehicles.

**To reproduce**
- Start a map created with vehicles
- Start a script using setModelHandling function.
- Only vehicles that have been just created will have new handling, but not the ones, that were created before the script.

**Expected behaviour**
If the vehicle model handling is changed then it would make sense for the handling of any already created vehicles to change.

**Screenshots**

**Version**
Unstated at mantis.

**Additional context**

From https://bugs.mtasa.com/view.php?id=7454

**Comment by Dutchman101** [comment]
I can confirm that the function doesn't affect already created vehicles, but I think this is only to be expected because setModelHandling has the positive trait of avoiding to modify handling of vehicles that may have already been modified by another script.

Examples:
1) resource spawns vehicle and immediately modifies its handling
2) resource or player spawns vehicle and player modifies handling in a "tune shop" or with a handling editor

To complement that, currently setModelHandling also doesn't override handling changes that were made after initially spawning a vehicle that is modified using setModelHandling as well.

Example:
1) vehicle gets spawned, correctly uses the customized handling from setModelHandling
- player or resource modifies said customized handling (changes from setModelHandling and this will co-exist) and setModelHandling won't undo these changes. This means setModelHandling isn't to be used to "hard code" handlings so that other scripts can't modify it, that was never the function's intention.

Perhaps if we check whether or not "already created vehicles" still have the default GTA handling (e.g were changed by another script) and mark them to determine if setModelHandling should apply retroactively, it would be possible to satisfy everyone. But personally I don't think it's worth it, or the cost of processing these checks for (and enumerating) all vehicles that were already created before a script using setModelHandling starts.

It would be far sought, especially because as I described in my examples, the way it currently works should be expected and is ideal to avoid interference of setModelHandling with other scripts. I want to close this issue, unless someone else has thoughts on it?

**Comment by Dutchman101** [comment]
> I want to close this issue, unless someone else has thoughts on it

**Comment by PlatinMTA** [comment]
> > I want to close this issue, unless someone else has thoughts on it

Hi. Maybe it would be nice to add an extra argument to setModelHandling so you would have the liberty to choose if you want to change the handling of already created vehicles with the same ID or no. Why? Because the alternative is to make a for loop and I don't think it's really necessary.

In my case one of the reasons I use this function is to avoid using setVehicleHandling at all, and on startup some vehicles use the old handling, then the new one. This is specially troublesome with multiple resources (as in my case). I can solve this by looping thourght all the created vehicles, getting their IDs and then setting the new handling one by one, but as I already said I don't think this should be necessary.

----

If this gets implemented it should look something like this:

```lua
bool setModelHandling ( int modelId, string property, var value [, bool existing = false ] )
```

existing: change already existing vehicles 

I can create a new issue if needed.

**Comment by patrikjuvonen** [comment]
I agree a new argument would be good. It's not very convenient to have to work around MTA's behavior in Lua.

**Comment by LosFaul** [comment]
> ```lua
> bool setModelHandling ( int modelId, string property, var value [, bool existing = false ] )
> ```

i personally would call it: overrideExisting instead of just existing
makes it a little bit more clear ;)

**Comment by redeibarnabas** [comment]
I support this.

Edit: also would call it overrideExisting.

**Comment by prnxdev** [comment]
I don't think it is good idea. There is `setVehicleHandling` function which allows to set handling for specific vehicle. That means developer can create loop over all vehicles with specific model and set handling. With multithreading (which is WIP) it will also be more optimized if we have too much vehicles in resource / map.

**Comment by prnxdev** [comment]
There is no problem also to create some static object in Lua with defined handlings for models and after resource start event set those values for specific models. This could be used also for creating new vehicle. Just after `createVehicle` use `setVehicleHandling`

```lua
local modelHandlings = {
   456 = {
      -- key, value object for specific model
   }
}

addEventHandler("onResourceStart", root, function () 
   for model, modelHandling in ipairs(modelHandlings) do
      for key, value in pairs(handling) do
         setModelHandling(model, key, value)
      end 
   end
end)

-- later in code
--- x, y, z are defined earlier
local model = 465
local vehicle = createVehicle(model, x, y, z)
local modelHandling = modelHandlings[model]

for key, value in pairs(handling) do
   setVehicleHandling(model, key, value)
end
```

**Comment by redeibarnabas** [comment]
Why aren't you supporting something that could only be benefited from? The option is false by default, no resource could break, and the whole point of this is to not have to loop through every vehicle, so your solution is irrelevant.

**Comment by prnxdev** [comment]
> Why aren't you supporting something that could only be benefited from?

I think we should focus on fixing bugs and creating new features instead of adding some stuff that can be easly implemented using current Lua functions and methods just because someone is lazy (?) and don't want to write things on his own (for me it looks like this). 

> whole point of this is to not have to loop through every vehicle, so your solution is irrelevant.

I didn't show you example with loop over all vehicles.

Just think about it in other way: your solution will add to MTA some new mechanism which will set vehicle handling when this element is created. It means that more resources will be used because there should be created new thread to watch element tree changes. Many servers don't actually change handling but this type of mechanism will start no matter if you do want to change this in future or not. It means also that server will consume more RAM and CPU. That is also why I don't support that.

**Comment by redeibarnabas** [comment]
It isn't gonna use more resources. Only when calling the function and the overrideExisting is set to true. Nothing else will change.

Second of all, this whole project is open-source, everyone can decide what they wanna work on. If you do not support this just because it isn't a priority, you shouldn't comment that you aren't supporting it, cause you are gonna deter people from developing it.

**Comment by prnxdev** [comment]
Yes it is open source but somehow you cannot push directly to master branch and publish it. Also someone have to accept your code changes before official release so it's not so "open for everyone" project - only source code is free for everyone. 

I just don't support things tha can be done with existing functions. 

Also there is one big problem with MTA right now: it has too many functions that can be easly implemented without any changes inside source code (like `Account` functions which are actually functions that are executing SQL statements to SQLite) or by just unlocking few Lua functions.

I support many things that are cool to use in future like Threading and things like that cannot be implemented using actually MTA API.

**Comment by redeibarnabas** [comment]
Why is it a problem for you that MTA makes it easier for low-level developers to actually be able to do something?

**Comment by AlexTMjugador** [comment]
> I don't think it is good idea. There is `setVehicleHandling` function which allows to set handling for specific vehicle. That means developer can create loop over all vehicles with specific model and set handling. With multithreading (which is WIP) it will also be more optimized if we have too much vehicles in resource / map.

Multithreading is not a silver bullet that you can throw in any loop that does the same thing for every element of a collection and always results in increased performance. In this case, the client will have to serialize setting the handling on the GTA thread, which means that it is impossible to set the handling of two vehicles at the same time and there will be no performance benefit there, but a performance hit (due to the extra operations needed for thread safety). On the server, maybe there is a bit of room for improvement by generating and sending several handling packets concurrently, but I wouldn't count on that providing the linear speedup we all dream about, and that should be transparent to scripts anyway.

Although I'm usually against bloating MTA API with functions that do not help its completeness, this is not even a new function, just an extra parameter to change a behavior that may be awkward for some valid use cases. I agree with @patrikjuvonen that MTA should not get in the way of people using its Lua API in ways that are arguably reasonable, and working around that in Lua, when it barely adds any maintenance burden to MTA codebase, is just bad design IMO.

**Comment by patrikjuvonen** [comment]
I didn't even know it didn't work on existing vehicles. I think this is a must-have, as it is a very common need for new handling to apply to all vehicles of the same model.
